File: safesock/cpp/examples/matrix_client.cpp
#include "../SafeSocket.hpp"
#include <iostream>
#include <string>
#include <vector>
#include <chrono>

int main(int argc, char* argv[]) {
    std::string address = "127.0.0.1:9999";
    if (argc > 1) {
        address = argv[1];
    }

    std::cout << "C++ Client: Connecting to " << address << std::endl;

    try {
        std::string public_ip = "1.2.3.4-cpp";
        safesock::SocketConfig config;
        config.public_ip = public_ip;

        // Create client with tcp-hello profile and custom identity
        auto client = safesock::create_with_config("tcp-hello:cpp-matrix", address, config, "client", true);
        std::cout << "C++ Client: Connected" << std::endl;

        // --- TEST 1: Basic Ping ---
        std::string payload = "ping-cpp";
        std::cout << "C++ Client: Sending " << payload << std::endl;
        std::vector<uint8_t> data(payload.begin(), payload.end());
        client->send(data);

        auto response = client->receive();
        std::string decoded(response.begin(), response.end());
        if (decoded != "echo:" + payload) {
             std::cerr << "C++ Client: FAILURE - Basic ping failed. Got: " << decoded << std::endl;
             return 1;
        }

        // --- TEST 2: Metadata Verification ---
        std::cout << "C++ Client: Verifying metadata..." << std::endl;
        std::string meta_req = "meta_request";
        std::vector<uint8_t> meta_data(meta_req.begin(), meta_req.end());
        client->send(meta_data);
        
        auto meta_resp_vec = client->receive();
        std::string meta_resp(meta_resp_vec.begin(), meta_resp_vec.end());
        
        if (meta_resp.find("meta:cpp-matrix") == std::string::npos || meta_resp.find(public_ip) == std::string::npos) {
            std::cerr << "C++ Client: FAILURE - Metadata mismatch. Got: " << meta_resp << std::endl;
            return 1;
        }
        std::cout << "C++ Client: Metadata verified" << std::endl;

        // --- TEST 3: Large Payload (1MB) ---
        std::cout << "C++ Client: Testing 1MB payload..." << std::endl;
        std::vector<uint8_t> large_payload(1024 * 1024, 'C');
        auto start = std::chrono::high_resolution_clock::now();
        client->send(large_payload);
        
        auto large_resp = client->receive(1024 * 1024 + 10);
        auto end = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();

        if (large_resp.size() != large_payload.size() + 5) {
            std::cerr << "C++ Client: FAILURE - Large payload length mismatch. Expected " << large_payload.size() + 5 << " but got " << large_resp.size() << std::endl;
            return 1;
        }
        
        std::cout << "C++ Client: 1MB payload verified in " << duration << "ms" << std::endl;

        std::cout << "C++ Client: ALL TESTS SUCCESS" << std::endl;
        return 0;

    } catch (const safesock::SafeSocketError& e) {
        std::cerr << "C++ Client: ERROR - " << e.what() << std::endl;
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "C++ Client: Unexpected error - " << e.what() << std::endl;
        return 1;
    }
}

File: safesock/rust/examples/matrix_client.rs
use safesocket::{SafeSocket, SocketConfig};
use std::env;
use std::process;
use std::time::Instant;

fn main() {
    let args: Vec<String> = env::args().collect();
    let addr = if args.len() > 1 { &args[1] } else { "127.0.0.1:9999" };

    let lib_name = if cfg!(target_os = "macos") {
        "libsafesocket.dylib"
    } else if cfg!(target_os = "windows") {
        "libsafesocket.dll"
    } else {
        "libsafesocket.so"
    };

    let possible_paths = vec![
        format!("./safesock/libsafesocket/{}", lib_name),
        format!("../libsafesocket/{}", lib_name),
        format!("../../libsafesocket/{}", lib_name),
        format!("./libsafesocket/{}", lib_name),
    ];

    let mut lib_path = String::new();
    for path in possible_paths {
        if std::path::Path::new(&path).exists() {
            lib_path = path;
            break;
        }
    }

    if lib_path.is_empty() {
        eprintln!("Rust Client: ERROR - Could not find libsafesocket");
        process::exit(1);
    }

    println!("Rust Client: Connecting to {}", addr);

    let public_ip = "1.2.3.4-rust";
    let config = SocketConfig {
        public_ip: public_ip.to_string(),
        ..Default::default()
    };

    match SafeSocket::new("tcp-hello:rust-matrix", addr, Some(config), "client", true, &lib_path) {
        Ok(sock) => {
            println!("Rust Client: Connected");
            
            // --- TEST 1: Basic Ping ---
            let payload = "ping-rust";
            sock.send(payload.as_bytes()).expect("Send failed");
            let response = sock.receive(1024).expect("Receive failed");
            let decoded = String::from_utf8_lossy(&response);
            if decoded != format!("echo:{}", payload) {
                eprintln!("Rust Client: FAILURE - Basic ping failed. Got: {}", decoded);
                process::exit(1);
            }

            // --- TEST 2: Metadata Verification ---
            println!("Rust Client: Verifying metadata...");
            sock.send(b"meta_request").expect("Send failed");
            let response = sock.receive(1024).expect("Receive failed");
            let decoded = String::from_utf8_lossy(&response);
            if !decoded.starts_with("meta:") {
                eprintln!("Rust Client: FAILURE - Metadata request failed");
                process::exit(1);
            }
            
            let parts: Vec<&str> = decoded[5..].split(',').collect();
            if parts[0] != "rust-matrix" || parts[2] != public_ip {
                eprintln!("Rust Client: FAILURE - Metadata mismatch. Expected name=rust-matrix, ip={}. Got: {}", public_ip, decoded);
                process::exit(1);
            }
            println!("Rust Client: Metadata verified (Host reported: {})", parts[1]);

            // --- TEST 3: Large Payload (1MB) ---
            println!("Rust Client: Testing 1MB payload...");
            let large_payload = vec![b'R'; 1024 * 1024];
            let start = Instant::now();
            sock.send(&large_payload).expect("Send failed");
            
            let response = sock.receive(1024 * 1024 + 10).expect("Receive failed");
            let duration = start.elapsed();
            
            if response.len() != large_payload.len() + 5 || !response.starts_with(b"echo:") {
                eprintln!("Rust Client: FAILURE - Large payload corrupted. Length={}", response.len());
                process::exit(1);
            }
            
            println!("Rust Client: 1MB payload verified in {:?}", duration);
            
            println!("Rust Client: ALL TESTS SUCCESS");
            process::exit(0);
        }
        Err(e) => {
            eprintln!("Rust Client: ERROR - Failed to create socket: {}", e);
            process::exit(1);
        }
    }
}
